src/tactic/ext.lean

```diff
@@ -375,23 +375,58 @@ do tgt ← target >>= whnf,
      then rintro [x] >> try_intros xs
      else pure (x :: xs)
 
+/-- Helper structure for `ext` and `ext1`. `lemmas` keeps track of extensionality lemmas
+  applied so far. -/
+meta structure ext_state : Type :=
+(patts : list rcases_patt := [])
+(lemmas : list name := [])
+(fuel : option ℕ := none)
+
+/-- Apply one extensionality lemma, and destruct the arguments using the patterns
+  in the ext_state. -/
+meta def ext1_core (cfg : apply_cfg := {}) : state_t ext_state tactic unit :=
+do ⟨patts, lemmas, fuel⟩ ← get,
+   (new_patts, new_lemmas) ← state_t.lift $ focus1 $
+   do { m ← get_ext_lemmas,
+         subject ← (target >>= get_ext_subject),
+         new_lemmas ←
+           do { rule ← (m.find subject),
+                (applyc rule cfg),
+                pure (lemmas ++ [rule]) } <|>
+             do { ls ← get_ext_lemma_names,
+                  let nms := ls.map name.to_string,
+                  rule ← (ls.any_of (λ n, applyc n cfg *> pure n)),
+                  pure (lemmas ++ [rule]) } <|>
+               (fail format!"no applicable extensionality rule found for {subject}"),
+         new_patts ← try_intros patts,
+         pure (new_patts, new_lemmas) },
+    modify (λ ⟨_, _, fuel⟩, ⟨new_patts, new_lemmas, fuel⟩)
+
+/-- Apply multiple extensionality lemmas, destructing the arguments using the given patterns. -/
+meta def ext_core (cfg : apply_cfg := {}) : state_t ext_state tactic unit :=
+do acc@⟨_, _, fuel⟩ ← get,
+   match fuel with
+   | (some 0) := pure ()
+   | n        := do { ext1_core cfg,
+                      modify (λ ⟨patts, lemmas, _⟩, ⟨patts, lemmas, nat.pred <$> n⟩),
+                      ext_core <|> pure () }
+   end
+
 /-- Apply one extensionality lemma, and destruct the arguments using the given patterns.
   Returns the unused patterns. -/
-meta def ext1 (xs : list rcases_patt) (cfg : apply_cfg := {}) : tactic (list rcases_patt) :=
-do subject ← target >>= get_ext_subject,
-   m ← get_ext_lemmas,
-   do { rule ← m.find subject,
-        applyc rule cfg } <|>
-     do { ls ← get_ext_lemma_names,
-          ls.any_of (λ n, applyc n cfg) } <|>
-     fail format!"no applicable extensionality rule found for {subject}",
-   try_intros xs
+meta def ext1 (xs : list rcases_patt) (cfg : apply_cfg := {})
+  (trace : bool := ff) : tactic (list rcases_patt) :=
+do ⟨_, σ⟩ ← state_t.run (ext1_core cfg) {patts := xs},
+   when trace (σ.lemmas.mmap' tactic.trace),
```

Comment:
Correct me if I'm wrong, but right now this just traces out the names of the `ext` lemmas that are used? Since all the other `?` versions of tactics output a `Try this:`, should `ext?` do something similar? So something similar to what we get with `show_term { ext }`

(btw I haven't looked into how hard it'd be)

Reply:
done

it's not difficult, although I agree with Simon's comment in issue #1481 that one would not want to inline `ext` as much as, say, `tidy`